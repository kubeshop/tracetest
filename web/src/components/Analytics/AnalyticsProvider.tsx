import {ReactElement, useEffect, useState, createContext, useCallback, useMemo} from 'react';
import {isEnabled, instance} from '../../services/Analytics/Analytics.service';

export const Context = createContext({
  isEnabled,
  instance,
});

type TAnalyticsProviderProps = {
  children: ReactElement;
};

const AnalyticsProvider: React.FC<TAnalyticsProviderProps> = ({children}) => {
  const [isInitialized, setInitialized] = useState(!isEnabled);

  const initialize = useCallback(() => {
    instance.initialize().finally(() => setInitialized(true));
  }, []);

  useEffect(() => {
    if (isEnabled) initialize();
  }, [initialize]);

  const value = useMemo(() => ({isEnabled, instance}), []);

  return isInitialized ? <Context.Provider value={value}>{children}</Context.Provider> : <div />;
};

export default AnalyticsProvider;
