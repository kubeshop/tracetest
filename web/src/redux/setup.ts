import {Middleware} from '@reduxjs/toolkit';

import OtelRepoAPI from 'redux/apis/OtelRepo';
import TestSpecs from 'redux/slices/TestSpecs.slice';
import Spans from 'redux/slices/Span.slice';
import DAG from 'redux/slices/DAG.slice';
import Trace from 'redux/slices/Trace.slice';
import CreateTestSuite from 'redux/slices/CreateTestSuite.slice';
import User from 'redux/slices/User.slice';
import TestOutputs from 'redux/testOutputs/slice';

export const middlewares: Middleware[] = [OtelRepoAPI.middleware];

export const reducers = {
  [OtelRepoAPI.reducerPath]: OtelRepoAPI.reducer,

  spans: Spans,
  dag: DAG,
  trace: Trace,
  testSpecs: TestSpecs,
  createTestSuite: CreateTestSuite,
  user: User,
  testOutputs: TestOutputs,
};
