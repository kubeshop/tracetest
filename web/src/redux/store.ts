import {configureStore, ThunkAction, Action} from '@reduxjs/toolkit';
import TestAPI from 'redux/apis/Test.api';
import ResultListSlice from 'redux/slices/ResultList.slice';

export const store = configureStore({
  reducer: {
    [TestAPI.reducerPath]: TestAPI.reducer,
    resultList: ResultListSlice,
  },
  middleware: getDefaultMiddleware => getDefaultMiddleware().concat(TestAPI.middleware),
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, RootState, unknown, Action<string>>;
