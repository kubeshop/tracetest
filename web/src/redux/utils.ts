import {PromiseWithKnownReason} from '@reduxjs/toolkit/dist/query/core/buildMiddleware/types';
import webSocketGateway, {IListenerFunction} from 'gateways/WebSocket.gateway';

export type {IListenerFunction} from 'gateways/WebSocket.gateway';

interface IInitWebSocketSubscription {
  listener: IListenerFunction;
  resource: string;
  waitToCleanSubscription: Promise<void>;
  waitToInitSubscription: PromiseWithKnownReason<any, any>;
}

/**
 * Implements RTK Query logic to receive streaming updates for persistent queries.
 * It allows a query to establish a WebSocket subscription, and apply updates to
 * the cached data when the information is received from the server.
 * For more information please take a look: https://redux-toolkit.js.org/rtk-query/usage/streaming-updates
 */
export async function initWebSocketSubscription({
  listener,
  resource,
  waitToCleanSubscription,
  waitToInitSubscription,
}: IInitWebSocketSubscription) {
  try {
    await waitToInitSubscription;
    webSocketGateway.subscribe(resource, listener);
  } catch {
    // no-op in case `waitToCleanSubscription` resolves before `waitToInitSubscription`,
    // in which case `waitToInitSubscription` will throw
  }
  await waitToCleanSubscription;
  webSocketGateway.unsubscribe(resource, '1234');
}
