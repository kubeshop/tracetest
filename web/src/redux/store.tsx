import {Action, configureStore, ThunkAction} from '@reduxjs/toolkit';
import {createReduxHistoryContext} from 'redux-first-history';
import {createBrowserHistory} from 'history';
import TracetestAPI from 'redux/apis/Tracetest';

import RouterMiddleware from './Router.middleware';
import {middlewares, reducers} from './setup';

const {createReduxHistory, routerMiddleware, routerReducer} = createReduxHistoryContext({
  history: createBrowserHistory(),
});

TracetestAPI.create();

export const store = configureStore({
  reducer: {
    ...reducers,
    [TracetestAPI.instance.reducerPath]: TracetestAPI.instance.reducer,
    router: routerReducer,
  },
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware()
      .prepend(RouterMiddleware.middleware)
      .concat(TracetestAPI.instance.middleware, ...middlewares, routerMiddleware),
});

export const history = createReduxHistory(store);

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, RootState, unknown, Action<string>>;
