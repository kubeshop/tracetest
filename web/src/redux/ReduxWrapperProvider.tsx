// test-utils.jsx
import * as React from 'react';
import {Provider} from 'react-redux';
import {store} from './store';

export const ReduxWrapperProvider: React.FC = ({children}) => {
  return <Provider store={store}>{children}</Provider>;
};
