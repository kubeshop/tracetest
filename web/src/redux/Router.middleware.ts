import {createListenerMiddleware} from '@reduxjs/toolkit';
import type {TypedStartListening} from '@reduxjs/toolkit';
import {LOCATION_CHANGE} from 'redux-first-history';
import {parse} from 'query-string';
import RouterActions from './actions/Router.actions';
import {runOriginPathAdded} from './slices/User.slice';
import {AppDispatch, RootState} from './store';

type AppStartListening = TypedStartListening<RootState, AppDispatch>;
const listener = createListenerMiddleware();
const startAppListening = listener.startListening as AppStartListening;

const sideEffectActionList = [RouterActions.updateSelectedAssertion, RouterActions.updateSelectedSpan];

const runUrlRegex = /^\/(test|testsuite)\/([^\/]+)\/run\/([^\/]+)(.*)$/;

const RouterMiddleware = () => ({
  middleware: listener.middleware,

  startListening(params = {}) {
    return startAppListening({
      predicate: ({type = ''}) =>
        type === LOCATION_CHANGE || (type.endsWith('/fulfilled') && !type.startsWith('router/')),
      effect(_, {dispatch, getState}) {
        const {
          router: {location},
        } = getState();

        const search = parse(location?.search || '');

        sideEffectActionList.forEach(action => {
          dispatch(action({search, params}));
        });
      },
    });
  },

  startListeningForLocationChange() {
    return startAppListening({
      predicate: action => {
        const pathname = action?.payload?.location?.pathname ?? '';
        return action?.type === LOCATION_CHANGE && pathname.match(runUrlRegex);
      },
      effect: async (_, {dispatch, getOriginalState, getState}) => {
        const {
          router: {location: prevLocation},
        } = getOriginalState();

        const {
          router: {location: currLocation},
        } = getState();

        const prevPathname = prevLocation?.pathname ?? '';

        if (!prevPathname.match(runUrlRegex) && !prevPathname.includes('/create')) {
          const defaultPath = currLocation?.pathname?.includes('testsuite') ? '/testsuites' : '/tests';
          dispatch(runOriginPathAdded(prevLocation?.pathname ?? defaultPath));
        }
      },
    });
  },
});

export default RouterMiddleware();
