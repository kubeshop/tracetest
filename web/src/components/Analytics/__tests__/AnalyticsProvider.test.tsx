import {render} from '@testing-library/react';
import {Provider} from 'react-redux';
import {store} from '../../../redux/store';
import AnalyticsProvider from '../index';

test('AnalyticsProvider', () => {
  const {container, getAllByTestId} = render(
    <Provider store={store}>
      <AnalyticsProvider>
        <h2 data-cy="sample">Cesco</h2>
      </AnalyticsProvider>
    </Provider>
  );
  expect(container).toMatchSnapshot();
  expect(getAllByTestId('not_initialized').length).toBe(1);
});
