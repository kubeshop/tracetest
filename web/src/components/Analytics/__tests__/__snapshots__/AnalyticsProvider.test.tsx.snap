// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnalyticsProvider 1`] = `
<div>
  <div
    data-cy="not_initialized"
  />
</div>
`;
